{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "489b8510-2554-40c6-ab4c-b06e9174eaf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import skbio\n",
    "from skbio import TreeNode\n",
    "import dendropy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "71fe9298-4e52-4e16-9422-835f94df36bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "25213f67-e657-47df-a95a-71b4a0bcf716",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch.nn as nn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4fb81098-bca7-4a43-8de6-e757329ff95c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import biom\n",
    "from biom.util import biom_open\n",
    "import torch\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6c0fb269-c48a-4bbe-8e43-54ded56431b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from skbio import TreeNode\n",
    "from scipy.spatial.distance import squareform\n",
    "from scipy.cluster.hierarchy import linkage, fcluster, leaves_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5dbbb860-740f-4b87-8037-0116780019a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import KFold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e6f7b9e9-05bc-4e83-974f-6076015620e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0a72a784-47f3-4ab2-8ec5-be5b154821ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import LabelEncoder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "09b0fc04-c0ec-496d-86c7-b7db28e52e02",
   "metadata": {},
   "outputs": [],
   "source": [
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "effbad56-906b-4ce7-b07b-4d287caf29a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gaussian_kernel(d, rho=2):\n",
    "    \"\"\"\n",
    "    weight = exp(-2 * ρ * d^2)\n",
    "    \"\"\"\n",
    "    return np.exp(-2 * rho * d ** 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29f1b56c-1ff4-4929-8c87-ab5c315b5612",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_graph_from_sample(otu_list, abundance_list, label, sample_id, full_tree):\n",
    "\n",
    "    leaf_abundance_map = {otu: abund for otu, abund in zip(otu_list, abundance_list)}\n",
    "    \n",
    "    # tree_sample = full_tree.shear(otu_list)\n",
    "    tree_sample = full_tree\n",
    "    all_nodes_in_tree = list(tree_sample.traverse(self_before=True))\n",
    "    node_to_idx = {node.name: i for i, node in enumerate(all_nodes_in_tree)}\n",
    "\n",
    "    full_abundance_map = {}\n",
    "    x = torch.zeros((len(all_nodes_in_tree), 1), dtype=torch.float)\n",
    "    edge_list, edge_weight_list = [], []\n",
    "    node_id = []\n",
    "    for node in tree_sample.postorder():\n",
    "        node_id.append(node_to_idx[node.name])\n",
    "        if node.is_tip():\n",
    "            full_abundance_map[node.name] = leaf_abundance_map.get(node.name, 0.0)\n",
    "            x[node_to_idx[node.name]] = full_abundance_map[node.name]\n",
    "        else:\n",
    "            child_abundances = [full_abundance_map.get(child.name, 0.0) for child in node.children]\n",
    "            full_abundance_map[node.name] = sum(child_abundances)\n",
    "            x[node_to_idx[node.name]] = full_abundance_map[node.name]\n",
    "            \n",
    "        if not node.is_root():\n",
    "            node_idx = node_to_idx[node.name]\n",
    "            parent_idx = node_to_idx[node.parent.name]\n",
    "            parent_idx = node_to_idx[node.parent.name]\n",
    "            edge_list.extend([[parent_idx, node_idx], [node_idx, parent_idx]])\n",
    "            distance = node.length if node.length is not None else 0.0\n",
    "            weight = gaussian_kernel(distance, rho=5e-5) \n",
    "            edge_weight_list.extend([weight, weight])\n",
    "            \n",
    "    edge_index = torch.tensor(edge_list, dtype=torch.long).t().contiguous()\n",
    "    y = torch.tensor([label], dtype=torch.long)\n",
    "    x = torch.tensor(x, dtype=torch.float)\n",
    "    edge_attr = torch.tensor(edge_weight_list, dtype=torch.float).view(-1, 1)\n",
    "    node_id = torch.tensor(node_id, dtype=torch.long)\n",
    "    \n",
    "    data = Data(x=x, node_id=node_id, edge_index=edge_index, edge_attr=edge_attr, y=y, sample_id=sample_id)\n",
    "\n",
    "    return data, x, edge_index, edge_attr, node_to_idx"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17c9a829-224d-455a-8ae1-e5d1196fa997",
   "metadata": {},
   "source": [
    "### IBD 16S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f040506-9ea1-4ed5-bd30-3de61613c2e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\"../data/IBD_16S/metadata.tsv\", sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16e65a5f-65c4-43ec-85fb-41118872e719",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = biom.load_table(\"../data/IBD_16S/table.biom\")\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    table_train.remove_empty()\n",
    "    with biom.util.biom_open(f\"../data/IBD_16S/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    table_test.remove_empty()\n",
    "    with biom.util.biom_open(f\"../data/IBD_16S/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"../data/IBD_16S/data/Real_Dateset_16S_IBD/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"../data/IBD_16S/data/Real_Dateset_16S_IBD/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e984865f-98ce-4751-b61c-43f93f8c0bbc",
   "metadata": {},
   "source": [
    "### CRC 16S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1028,
   "id": "d4c308f7-fd1f-44c4-9e95-cb931f258fb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_csv(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/16S_CRC.csv\")\n",
    "\n",
    "metadata = pd.DataFrame({\"sample_id\": table.iloc[:, 0].values, \"group\": table[\"Group\"].values})\n",
    "metadata[\"group\"] = [0 if i == \"HC\" else 1 for i in metadata[\"group\"].values]\n",
    "metadata.to_csv(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/metadata.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "table = biom.Table(data=table.iloc[:, 1:-1].values.T, observation_ids=table.iloc[:, 1:-1].columns.values, sample_ids=table.iloc[:, 0].values)\n",
    "with biom.util.biom_open(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/table.biom\", 'w') as f:\n",
    "    table.to_hdf5(f, 'example')\n",
    "\n",
    "metadata = metadata.set_index(\"sample_id\")\n",
    "table = biom.load_table(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/table.biom\")\n",
    "\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "id": "90276f55-0673-4ed3-adfb-50c1a86f3966",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1, 6):\n",
    "    table_train = biom.load_table(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/data/train_{i}.biom\")\n",
    "    table_test = biom.load_table(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/data/test_{i}.biom\")\n",
    "    for m in [\"without_low\", \"without_high\"]:\n",
    "        for n in [0.1, 0.2, 0.4, 0.5, 0.6, 0.8]:\n",
    "            tree = TreeNode.read(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/phylogeny_{m}_{n}.nwk\")\n",
    "            leaf_names = [node.name for node in tree.tips()]\n",
    "            \n",
    "            table_train_pick = table_train.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/data/train_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_train_pick.to_hdf5(f, 'example')\n",
    "                \n",
    "            table_test_pick = table_test.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/data/test_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_test_pick.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "135a3a88-7ded-48bd-9a01-f70e072f2d95",
   "metadata": {},
   "source": [
    "### CRC WGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1031,
   "id": "48c65e26-b223-4bd9-97ab-9244576cf09f",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_csv(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/WGS_CRC.csv\", sep=\"\\t\")\n",
    "\n",
    "metadata = pd.DataFrame({\"sample_id\": table.iloc[:, 0].values, \"group\": table[\"Group\"].values})\n",
    "metadata[\"group\"] = [0 if i == \"CTR\" else 1 for i in metadata[\"group\"].values]\n",
    "metadata.to_csv(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/metadata.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "fid = table.iloc[:, 1:-1].columns.values.astype(str)\n",
    "fid = np.char.replace(fid, '__', '.')\n",
    "fid = np.char.replace(fid, '_', '.')\n",
    "table = biom.Table(data=table.iloc[:, 1:-1].values.T, observation_ids=fid, sample_ids=table.iloc[:, 0].values)\n",
    "with biom.util.biom_open(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/table.biom\", 'w') as f:\n",
    "    table.to_hdf5(f, 'example')\n",
    "\n",
    "metadata = metadata.set_index(\"sample_id\")\n",
    "table = biom.load_table(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/table.biom\")\n",
    "\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "id": "08de8a49-be9f-49be-9c49-ddb7b62a46ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1, 6):\n",
    "    table_train = biom.load_table(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/data/train_{i}.biom\")\n",
    "    table_test = biom.load_table(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/data/test_{i}.biom\")\n",
    "    for m in [\"without_low\", \"without_high\"]:\n",
    "        for n in [0.1, 0.2, 0.4, 0.5, 0.6, 0.8]:\n",
    "            tree = TreeNode.read(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/phylogeny_{m}_{n}.nwk\")\n",
    "            leaf_names = [node.name for node in tree.tips()]\n",
    "            \n",
    "            table_train_pick = table_train.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/data/train_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_train_pick.to_hdf5(f, 'example')\n",
    "                \n",
    "            table_test_pick = table_test.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/data/test_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_test_pick.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcd7f731-2290-4387-bdea-6bc453eba295",
   "metadata": {},
   "source": [
    "### T2D WGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1043,
   "id": "9e0d2673-e9c9-4124-b738-650ee8dda81a",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_csv(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/WGS_T2D.csv\")\n",
    "\n",
    "metadata = pd.DataFrame({\"sample_id\": table.iloc[:, 0].values, \"group\": table[\"Group\"].values})\n",
    "metadata[\"group\"] = [0 if i == \"Health\" else 1 for i in metadata[\"group\"].values]\n",
    "metadata.to_csv(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/metadata.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "fid = table.iloc[:, 1:-1].columns.values.astype(str)\n",
    "fid = np.char.replace(fid, '__', '.')\n",
    "fid = np.char.replace(fid, '_', '.')\n",
    "table = biom.Table(data=table.iloc[:, 1:-1].values.T, observation_ids=fid, sample_ids=table.iloc[:, 0].values)\n",
    "with biom.util.biom_open(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/table.biom\", 'w') as f:\n",
    "    table.to_hdf5(f, 'example')\n",
    "\n",
    "metadata = metadata.set_index(\"sample_id\")\n",
    "table = biom.load_table(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/table.biom\")\n",
    "\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "d62a0eca-e23e-4c2d-bae7-d711d95e445a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1, 6):\n",
    "    table_train = biom.load_table(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/data/train_{i}.biom\")\n",
    "    table_test = biom.load_table(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/data/test_{i}.biom\")\n",
    "    for m in [\"without_low\", \"without_high\"]:\n",
    "        for n in [0.1, 0.2, 0.4, 0.5, 0.6, 0.8]:\n",
    "            tree = TreeNode.read(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/phylogeny_{m}_{n}.nwk\")\n",
    "            leaf_names = [node.name for node in tree.tips()]\n",
    "            \n",
    "            table_train_pick = table_train.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/data/train_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_train_pick.to_hdf5(f, 'example')\n",
    "                \n",
    "            table_test_pick = table_test.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/data/test_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_test_pick.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0475759-9050-49a3-ac12-39429eb90fa2",
   "metadata": {},
   "source": [
    "### Dietary fiber"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "de5f1ebf-b120-455c-9557-6a3753055432",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6693 x 2270 <class 'biom.table.Table'> with 313231 nonzero entries (2% dense)"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_excel(\"/home/dongbiao/word_embedding_microbiome/programe_test/dietary_fber/data/metadata.xls\", index_col=0)\n",
    "table = biom.load_table(\"/home/dongbiao/word_embedding_microbiome/programe_test/dietary_fber/data/projects/table_6721_2378.biom\")\n",
    "metadata = metadata.loc[table.ids(axis=\"sample\")]\n",
    "metadata = metadata.loc[metadata.study.values != \"PRJEB2165\"]\n",
    "metadata = metadata.loc[metadata.study.values != \"PRJNA385004\"]\n",
    "table = table.filter(metadata.index.values)\n",
    "table.remove_empty()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "dbf937d5-2f43-4dea-b8cf-473d0d351032",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "930 x 2270 <class 'biom.table.Table'> with 252062 nonzero entries (11% dense)"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fid = table.ids(axis=\"observation\")\n",
    "prevalence = table.nonzero_counts(axis=\"observation\")\n",
    "table = table.filter(fid[prevalence > 50], axis='observation', inplace=False)\n",
    "table.remove_empty()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "574391be-e0b5-4c3d-9531-005dcb5f11cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = metadata.loc[table.ids(axis=\"sample\")]\n",
    "table = table.norm(axis='sample', inplace=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "83f44f70-3306-487e-9007-e07c9c513e64",
   "metadata": {},
   "outputs": [],
   "source": [
    "with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/table.biom\", 'w') as f:\n",
    "    table.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "88456b68-86fd-4fde-8cd7-5fefb3700b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"/home/dongbiao/GCN/data/dietary_fiber/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"/home/dongbiao/GCN/data/dietary_fiber/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "id": "387fc0fe-6ae6-4835-99e0-6663a28f54f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1, 6):\n",
    "    table_train = biom.load_table(f\"/home/dongbiao/GCN/data/dietary_fiber/data/train_{i}.biom\")\n",
    "    table_test = biom.load_table(f\"/home/dongbiao/GCN/data/dietary_fiber/data/test_{i}.biom\")\n",
    "    for m in [\"without_low\", \"without_high\"]:\n",
    "        for n in [0.1, 0.2, 0.4, 0.5, 0.6, 0.8]:\n",
    "            tree = TreeNode.read(f\"/home/dongbiao/GCN/data/dietary_fiber/phylogeny_{m}_{n}.nwk\")\n",
    "            leaf_names = [node.name for node in tree.tips()]\n",
    "            \n",
    "            table_train_pick = table_train.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/data/train_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_train_pick.to_hdf5(f, 'example')\n",
    "                \n",
    "            table_test_pick = table_test.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/data/test_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_test_pick.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "0b6d3339-38b5-4980-b905-db7fae0aee65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['PRJEB41443', 'PRJNA560950', 'PRJNA780023', 'PRJNA891951',\n",
       "       'PRJNA293971', 'PRJNA306884', 'PRJNA428736'], dtype=object)"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_id = metadata.study.unique()\n",
    "study_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "a1470384-cde3-42ad-93bc-6ea755e62c8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRJEB41443\n",
    "temp = metadata.loc[metadata.study == study_id[0]]\n",
    "sid_1 = temp.loc[temp.timepoint_numeric == 1].index.values\n",
    "sid_2 = temp.loc[temp.group == 1].index.values\n",
    "PRJEB41443_sid = list(sid_1) + list(sid_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "fecbb71f-636d-4dee-ba17-be33d3ffe819",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRJNA560950\n",
    "metadata.loc[metadata.study == study_id[1]].groupby([\"subject_id\", \"group\", \"timepoint_id\"]).count()\n",
    "temp = metadata.loc[metadata.study == study_id[1]]\n",
    "PRJNA560950_sid = temp.loc[temp.timepoint_id == \"W4\"].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "5ec859bc-514f-4ca9-8c8a-9c83b14d24b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRJNA780023\n",
    "temp = metadata.loc[metadata.study == study_id[2]]\n",
    "sid_1 = temp.loc[temp.group == 0].drop_duplicates(subset='subject_id').index.values\n",
    "sid_2 = temp.loc[temp.group == 1].drop_duplicates(subset='subject_id').index.values\n",
    "PRJNA780023_sid = list(sid_1) + list(sid_2) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "beedbc01-3a2c-4f67-834a-6ff7c905d117",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRJNA891951\n",
    "temp = metadata.loc[metadata.study == study_id[3]]\n",
    "PRJNA891951_sid = temp.loc[temp.timepoint == \"after\"].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "034cde1a-4dd4-4b5a-8314-4dc47afe7e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRJNA293971\n",
    "temp = metadata.loc[metadata.study == study_id[4]]\n",
    "PRJNA293971_sid = temp.loc[[i in [\"before_1\", \"after_3\"] for i in temp.timepoint_id.values]].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "6ce34010-78be-4f59-8d7d-a8fb2ba4eba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRJNA306884\n",
    "temp = metadata.loc[metadata.study == study_id[5]]\n",
    "PRJNA306884_sid = temp.loc[[i in [4, 8] for i in temp.timepoint_numeric.values]].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "bdb79875-af08-476b-85c7-055917e29af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRJNA428736\n",
    "temp = metadata.loc[metadata.study == study_id[6]]\n",
    "sid_1 = temp.loc[temp.group == 0].drop_duplicates(subset='subject_id').index.values\n",
    "sid_2 = temp.loc[temp.group == 1].drop_duplicates(subset='subject_id').index.values\n",
    "PRJNA428736_sid = list(sid_1) + list(sid_2) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "53241bab-6740-4aa5-ae9c-ad3eb9527d2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6235 x 731 <class 'biom.table.Table'> with 105646 nonzero entries (2% dense)"
      ]
     },
     "execution_count": 258,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pick_sid = PRJEB41443_sid + PRJNA560950_sid + PRJNA780023_sid + PRJNA891951_sid + PRJNA293971_sid + PRJNA306884_sid + PRJNA428736_sid\n",
    "table.filter(pick_sid, axis=\"sample\")\n",
    "table.remove_empty()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "1f00a690-53b9-49c4-aa14-ce6c59665691",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "936 x 731 <class 'biom.table.Table'> with 80180 nonzero entries (11% dense)"
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fid = table.ids(axis=\"observation\")\n",
    "prevalence = table.nonzero_counts(axis=\"observation\")\n",
    "table = table.filter(fid[prevalence > 20], axis='observation', inplace=False)\n",
    "table.remove_empty()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "2d5849dd-17d7-44c0-a5b2-7888d5af1293",
   "metadata": {},
   "outputs": [],
   "source": [
    "with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/table.biom\", 'w') as f:\n",
    "    table.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "3269ae7c-96f5-4a2e-b983-e9694e8e3c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = metadata.loc[table.ids(axis=\"sample\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "f6c02e7d-d023-4974-9f56-15a13cae3d7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = table.norm(axis='sample', inplace=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "c5f13248-dd6a-4948-816c-7312abf85f74",
   "metadata": {},
   "outputs": [],
   "source": [
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"/home/dongbiao/GCN/data/dietary_fiber/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"/home/dongbiao/GCN/data/dietary_fiber/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "1e34d47b-c3e9-429d-939f-1a988554effa",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1, 6):\n",
    "    table_train = biom.load_table(f\"/home/dongbiao/GCN/data/dietary_fiber/data/train_{i}.biom\")\n",
    "    table_test = biom.load_table(f\"/home/dongbiao/GCN/data/dietary_fiber/data/test_{i}.biom\")\n",
    "    for m in [\"without_low\", \"without_high\"]:\n",
    "        for n in [0.1, 0.2, 0.4, 0.5, 0.6, 0.8]:\n",
    "            tree = TreeNode.read(f\"/home/dongbiao/GCN/data/dietary_fiber/phylogeny_{m}_{n}.nwk\")\n",
    "            leaf_names = [node.name for node in tree.tips()]\n",
    "            \n",
    "            table_train_pick = table_train.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/data/train_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_train_pick.to_hdf5(f, 'example')\n",
    "                \n",
    "            table_test_pick = table_test.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/GCN/data/dietary_fiber/data/test_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_test_pick.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20b13641-4ebb-4350-865d-ade6f0908d8f",
   "metadata": {},
   "source": [
    "### OSCC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1046,
   "id": "10caabda-7d7b-430b-99d2-5deeeb60bd33",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = biom.load_table(\"/home/dongbiao/word_embedding_microbiome/all_data/oral/global-oral-microbiome/data/table_gg2_id99_24409_56564_v5.biom\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1047,
   "id": "c4e6cd1c-96e0-4547-b904-1e1ee7fc6cef",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\"/home/dongbiao/GCN/data/OSCC/metadata.txt\", sep=\"\\t\", index_col=0, low_memory=False)\n",
    "metadata = metadata.loc[metadata.diagnose != \"OPMD\"]\n",
    "metadata.loc[:, \"group\"] = [1 if i == \"OSCC\" else 0 for i in metadata.diagnose.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1048,
   "id": "12cc9bf2-8d36-4420-8ede-f79c93b73cb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "table_filter = None  \n",
    "for i, study_id in enumerate(metadata.project_ID.unique()):\n",
    "    metadata_study = metadata.loc[metadata.project_ID.values == study_id]\n",
    "    sid = np.intersect1d(metadata_study.index.values, table.ids(axis=\"sample\"))\n",
    "    temp = table.filter(sid, inplace=False)\n",
    "    temp.remove_empty()\n",
    "    \n",
    "    fid = temp.ids(axis=\"observation\")\n",
    "    prevalence = temp.nonzero_counts(axis=\"observation\")\n",
    "    temp = temp.filter(fid[prevalence > 10], axis='observation', inplace=False)\n",
    "    temp.remove_empty()\n",
    "    \n",
    "    if table_filter is None:\n",
    "        table_filter = temp\n",
    "    else:\n",
    "        table_filter = table_filter.merge(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1049,
   "id": "10c2c49c-bce6-4916-abfa-e759cab5b512",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "864 x 631 <class 'biom.table.Table'> with 88061 nonzero entries (16% dense)"
      ]
     },
     "execution_count": 1049,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1050,
   "id": "dcb7d2bf-fc35-4d70-a5a8-fe370aad463c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "569 x 631 <class 'biom.table.Table'> with 79776 nonzero entries (22% dense)"
      ]
     },
     "execution_count": 1050,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "otu_retained, otu_removed = filter_low_abundance_asvs(table_filter.norm(axis='sample', inplace=False).to_dataframe())\n",
    "table_filter = table_filter.filter(otu_retained, axis='observation', inplace=False)\n",
    "table_filter.remove_empty()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1051,
   "id": "9e308ee5-67be-4fa7-9b5c-1e4908249620",
   "metadata": {},
   "outputs": [],
   "source": [
    "with biom.util.biom_open(\"/home/dongbiao/GCN/data/OSCC/table.biom\", 'w') as f:\n",
    "    table_filter.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 862,
   "id": "9917b1f1-e03f-408c-9e28-2b433149ac03",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = metadata.loc[table_filter.ids(axis=\"sample\")]\n",
    "metadata.to_csv(\"/home/dongbiao/GCN/data/OSCC/metadata.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1052,
   "id": "0ea96ac4-c981-4b64-86ed-8f015eb43bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\"/home/dongbiao/GCN/data/OSCC/metadata.tsv\", sep=\"\\t\", index_col=0)\n",
    "table = biom.load_table(\"/home/dongbiao/GCN/data/OSCC/table.biom\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1053,
   "id": "85ca0bac-114d-4283-90f0-782a3fdcafac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# study_id = metadata.project_ID.unique()\n",
    "# n = 1\n",
    "# for i in study_id:\n",
    "#     sid = metadata.loc[metadata.project_ID != i].index.values\n",
    "#     table_train = table.filter(sid, axis=\"sample\", inplace=False)\n",
    "#     with biom.util.biom_open(f\"/home/dongbiao/GCN/data/OSCC/data/train_{n}.biom\", 'w') as f:\n",
    "#         table_train.to_hdf5(f, 'example')\n",
    "#     sid = metadata.loc[metadata.project_ID == i].index.values\n",
    "#     table_test = table.filter(sid, axis=\"sample\", inplace=False)\n",
    "#     with biom.util.biom_open(f\"/home/dongbiao/GCN/data/OSCC/data/test_{n}.biom\", 'w') as f:\n",
    "#         table_test.to_hdf5(f, 'example')\n",
    "\n",
    "#     table_train = table_train.norm(axis='sample', inplace=False).to_dataframe().T\n",
    "#     table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "#     table_train.to_csv(f\"/home/dongbiao/GCN/data/OSCC/data/train_{n}.csv\")\n",
    "#     table_test = table_test.norm(axis='sample', inplace=False).to_dataframe().T\n",
    "#     table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "#     table_test.to_csv(f\"/home/dongbiao/GCN/data/OSCC/data/test_{n}.csv\")\n",
    "    \n",
    "#     n = n + 1\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/GCN/data/OSCC/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/GCN/data/OSCC/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"/home/dongbiao/GCN/data/OSCC/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"/home/dongbiao/GCN/data/OSCC/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6bafa1a-2b35-4119-966b-2aa294a0f643",
   "metadata": {},
   "source": [
    "### Multi-status classsification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d99e4ab7-a7a3-406e-ad9e-512dbf9d473d",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_csv(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/Multi-classification.csv\")\n",
    "metadata = pd.DataFrame({\"sample_id\": table.iloc[:, 0].values, \"disease\": table[\"Group\"].values,\n",
    "                         \"group\": encoder.fit_transform(table[\"Group\"].values)})\n",
    "metadata.to_csv(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/metadata.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "fid = table.iloc[:, 1:-1].columns.values.astype(str)\n",
    "table = biom.Table(data=table.iloc[:, 1:-1].values.T, observation_ids=fid, sample_ids=table.iloc[:, 0].values)\n",
    "with biom.util.biom_open(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/table.biom\", 'w') as f:\n",
    "    table.to_hdf5(f, 'example')\n",
    "\n",
    "metadata = metadata.set_index(\"sample_id\")\n",
    "table = biom.load_table(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/table.biom\")\n",
    "\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "afb6f554-2cbf-43d3-b9c8-150505516623",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1, 6):\n",
    "    table_train = biom.load_table(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/data/train_{i}.biom\")\n",
    "    table_test = biom.load_table(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/data/test_{i}.biom\")\n",
    "    for m in [\"without_low\", \"without_high\"]:\n",
    "        for n in [0.1, 0.2, 0.4, 0.5, 0.6, 0.8]:\n",
    "            tree = TreeNode.read(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/phylogeny_{m}_{n}.nwk\")\n",
    "            leaf_names = [node.name for node in tree.tips()]\n",
    "            \n",
    "            table_train_pick = table_train.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/data/train_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_train_pick.to_hdf5(f, 'example')\n",
    "                \n",
    "            table_test_pick = table_test.filter(leaf_names, axis=\"observation\", inplace=False)\n",
    "            with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/data/test_{m}_{n}_{i}.biom\", 'w') as f:\n",
    "                table_test_pick.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0cbb65c-fdd0-4f40-841e-ea7ae9c120be",
   "metadata": {},
   "source": [
    "### Synthetic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "b9f01837-2c89-4673-a047-d5b9788d7e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GreengenesSyntheticGenerator:\n",
    "    def __init__(self, tree_path, n_features=60, random_seed=42):\n",
    "        \"\"\"\n",
    "        初始化生成器：基于真实的 Greengenes 树文件。\n",
    "        \"\"\"\n",
    "        self.n_features = n_features\n",
    "        self.rng = np.random.default_rng()\n",
    "        \n",
    "        print(f\"正在加载进化树: {tree_path} ...\")\n",
    "        self.full_tree = TreeNode.read(tree_path)\n",
    "        \n",
    "        # 随机选择 OTU\n",
    "        all_tips = [node.name for node in self.full_tree.tips()]\n",
    "            \n",
    "        self.selected_otus = self.rng.choice(all_tips, size=n_features, replace=False)\n",
    "        self.leaf_names = list(self.selected_otus)\n",
    "        \n",
    "        # 剪切树并计算距离\n",
    "        print(\"正在计算系统发育距离矩阵...\")\n",
    "        self.sheared_tree = self.full_tree.shear(self.leaf_names)\n",
    "        dm = self.sheared_tree.tip_tip_distances()\n",
    "        \n",
    "        self.phylo_dist_matrix = np.zeros((n_features, n_features))\n",
    "        for i, otu1 in enumerate(self.leaf_names):\n",
    "            for j, otu2 in enumerate(self.leaf_names):\n",
    "                self.phylo_dist_matrix[i, j] = dm[otu1, otu2]\n",
    "\n",
    "        # =================================================================\n",
    "        # 【修改重点】：强制平衡划分 Clade 1 和 Clade 2\n",
    "        # =================================================================\n",
    "        # 1. 使用 'ward' 方法进行聚类，它倾向于产生大小更均匀的簇（相比 average/single）\n",
    "        #    注意：虽然 UPGMA (average) 更符合生物学进化，但为了模拟平衡数据，Ward 更稳健\n",
    "        Z = linkage(squareform(self.phylo_dist_matrix), method='ward')\n",
    "        \n",
    "        # 2. 获取按树结构排序后的叶子节点索引列表\n",
    "        #    leaves_list 返回的顺序保证了相邻的索引在树上是亲缘关系最近的\n",
    "        ordered_indices = leaves_list(Z)\n",
    "        \n",
    "        # 3. 强制一分为二 (例如 60 -> 30 + 30)\n",
    "        #    这样既保证了数量平衡，又因为基于树排序，保证了组内的系统发育相似性\n",
    "        midpoint = len(ordered_indices) // 2\n",
    "        clade1_indices = ordered_indices[:midpoint]\n",
    "        clade2_indices = ordered_indices[midpoint:]\n",
    "        \n",
    "        # 4. 生成标签数组\n",
    "        self.cluster_labels = np.zeros(n_features, dtype=int)\n",
    "        self.cluster_labels[clade1_indices] = 1\n",
    "        self.cluster_labels[clade2_indices] = 2\n",
    "        \n",
    "        print(f\"聚类完成: Clade 1 ({len(clade1_indices)} 个), Clade 2 ({len(clade2_indices)} 个)\")\n",
    "\n",
    "    def _generate_base_abundance(self, n_samples, pattern_type):\n",
    "        data = np.zeros((n_samples, self.n_features))\n",
    "        \n",
    "        clade1_idx = np.where(self.cluster_labels == 1)[0]\n",
    "        clade2_idx = np.where(self.cluster_labels == 2)[0]\n",
    "        \n",
    "        if pattern_type == 'S1':\n",
    "            # S1 主要在 Clade 1 (Cluster 1)\n",
    "            active_idx = clade1_idx\n",
    "            # if len(clade2_idx) > 0:\n",
    "            #     active_idx = np.concatenate([active_idx, self.rng.choice(clade2_idx, size=int(len(clade2_idx)*0.1))])\n",
    "        else:\n",
    "            # S2 主要在 Clade 2 (Cluster 2)\n",
    "            active_idx = clade2_idx\n",
    "            # if len(clade1_idx) > 0:\n",
    "            #     active_idx = np.concatenate([active_idx, self.rng.choice(clade1_idx, size=int(len(clade1_idx)*0.1))])\n",
    "            \n",
    "        for i in range(n_samples):\n",
    "            base_vals = self.rng.dirichlet(np.ones(len(active_idx)), size=1)[0]\n",
    "            data[i, active_idx] = base_vals\n",
    "        return data\n",
    "\n",
    "    def generate_dataset_1(self):\n",
    "        \"\"\"\n",
    "        生成 Synthetic data set 1\n",
    "        修改说明：\n",
    "        - S3 (从 S1 扰动而来): 被扰动的特征从 S2 的主要特征 (Clade 2) 中选取。\n",
    "        - S4 (从 S2 扰动而来): 被扰动的特征从 S1 的主要特征 (Clade 1) 中选取。\n",
    "        \"\"\"\n",
    "        n_total_group1 = 77\n",
    "        n_total_group2 = 71\n",
    "        \n",
    "        # 1. 生成基础数据\n",
    "        data_g1 = self._generate_base_abundance(n_total_group1, 'S1')\n",
    "        data_g2 = self._generate_base_abundance(n_total_group2, 'S2')\n",
    "        \n",
    "        labels_g1 = np.zeros(n_total_group1)\n",
    "        labels_g2 = np.ones(n_total_group2)\n",
    "        \n",
    "        # 2. 添加噪声\n",
    "        data_g1 += self.rng.normal(0, 1e-6, data_g1.shape)\n",
    "        data_g2 += self.rng.normal(0, 1e-6, data_g2.shape)\n",
    "        \n",
    "        # 获取 S1 (Clade 1) 和 S2 (Clade 2) 对应的特征索引\n",
    "        # self.cluster_labels 在 __init__ 中已经生成\n",
    "        clade1_idx = np.where(self.cluster_labels == 1)[0]\n",
    "        clade2_idx = np.where(self.cluster_labels == 2)[0]\n",
    "\n",
    "        # 3. 引入特征错位 (Feature Misalignment)\n",
    "        # 修改：增加 target_pool 参数，用于指定从哪个集合中选取要被扰动的特征\n",
    "        def apply_misalignment(data, target_pool, move_id):\n",
    "            n_samples = data.shape[0]\n",
    "            # 随机选择 20%-30% 的样本进行扰动 (根据你的上一段代码逻辑调整回论文常见比例，或者保持你设定的0.8)\n",
    "            # 这里我保持你代码中的 0.8-0.9，如果需要论文原比例请改为 0.2-0.3\n",
    "            perturb_ratio = 0.90\n",
    "            n_perturb = int(n_samples * perturb_ratio)\n",
    "            perturb_indices = self.rng.choice(n_samples, n_perturb, replace=False)\n",
    "\n",
    "            for idx in perturb_indices:\n",
    "                # 随机选择一定比例的 OTU 进行移动\n",
    "                otu_ratio = self.rng.uniform(0.45, 0.45)\n",
    "                # 计算需要移动的特征数量\n",
    "                actual_n_move = int(len(move_id) * otu_ratio)\n",
    "                source_otus = self.rng.choice(move_id, actual_n_move, replace=False)\n",
    "                replace_otus = self.rng.choice(target_pool, actual_n_move, replace=False)\n",
    "\n",
    "                data[idx, replace_otus] = data[idx, source_otus]\n",
    "                data[idx, source_otus] = 0\n",
    "\n",
    "            return data\n",
    "\n",
    "        # S1 -> S3: 扰动 S2 的特征 (即从 clade2_idx 中选特征来捣乱)\n",
    "        # 逻辑：S1 样本中本来 S2 特征很少，现在我们特意去动这些特征(如果存在的话)\n",
    "        data_g1 = apply_misalignment(data_g1, target_pool=clade2_idx, move_id=clade1_idx)\n",
    "        \n",
    "        # S2 -> S4: 扰动 S1 的特征 (即从 clade1_idx 中选特征来捣乱)\n",
    "        data_g2 = apply_misalignment(data_g2, target_pool=clade1_idx, move_id=clade2_idx)\n",
    "        \n",
    "        # 非负截断\n",
    "        data_g1 = np.maximum(data_g1, 0)\n",
    "        data_g2 = np.maximum(data_g2, 0)\n",
    "        \n",
    "        X = np.vstack([data_g1, data_g2])\n",
    "        y = np.concatenate([labels_g1, labels_g2])\n",
    "        \n",
    "        return X, y, self.leaf_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 604,
   "id": "11d9ab48-0cf4-469f-bfc6-6f6b0b7858d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "正在加载进化树: /home/dongbiao/GCN/data/synthetic_data/phylogeny.nwk ...\n",
      "正在计算系统发育距离矩阵...\n",
      "聚类完成: Clade 1 (100 个), Clade 2 (100 个)\n"
     ]
    }
   ],
   "source": [
    "tree_path = \"/home/dongbiao/GCN/data/synthetic_data/phylogeny.nwk\"\n",
    "\n",
    "generator = GreengenesSyntheticGenerator(tree_path, n_features=200)\n",
    "\n",
    "X, y, otu_ids = generator.generate_dataset_1()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "id": "0b826ab0-fac3-463f-9707-f24ebbb27d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "s = [f\"S_{i}\" for i in range(1, X.shape[0]+1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "id": "f93a766d-859f-4003-a3fe-eb5bad5c7d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = biom.Table(data=X.T, observation_ids=otu_ids, sample_ids=s)\n",
    "with biom.util.biom_open(\"/home/dongbiao/GCN/data/synthetic_data/table.biom\", 'w') as f:\n",
    "    table.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "id": "ae1818f2-8cd7-4c39-ba45-72273ef0a5f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.DataFrame({\"sample_id\": s, \"group\": y.astype(np.int32)})\n",
    "metadata.to_csv(\"/home/dongbiao/GCN/data/synthetic_data/metadata.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 608,
   "id": "9147e39a-7fad-4a1b-9090-9cc06021e10e",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = metadata.set_index(\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 609,
   "id": "25a4d3dd-9a4f-42c0-91cf-b0a89a669145",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = biom.load_table(\"/home/dongbiao/GCN/data/synthetic_data/table.biom\")\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    table_train.remove_empty()\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/GCN/data/synthetic_data/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    table_test.remove_empty()\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/GCN/data/synthetic_data/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"/home/dongbiao/GCN/data/synthetic_data/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"/home/dongbiao/GCN/data/synthetic_data/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d17c254-bac9-4028-9fa1-fa8d2f3630f8",
   "metadata": {},
   "source": [
    "### Synthetic data suxiaoquan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 641,
   "id": "65c5273c-43be-4c97-9b2c-938ac297a264",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_csv(\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/Synthetic_Dataset_1.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 643,
   "id": "d35a644e-6ba7-42aa-a4b7-40bbd929e811",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.DataFrame({\"sample_id\": table.iloc[:, 0].values, \"group\": table[\"Group\"].values})\n",
    "metadata[\"group\"] = [0 if i == \"Group_S1_S3\" else 1 for i in metadata[\"group\"].values]\n",
    "metadata.to_csv(\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/metadata.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 635,
   "id": "77895e8a-9a0a-4aef-8c7b-2cb75c39409e",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = biom.Table(data=table.iloc[:, 1:61].values.T, observation_ids=table.iloc[:, 1:61].columns.values, sample_ids=table.iloc[:, 0].values)\n",
    "with biom.util.biom_open(\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/table.biom\", 'w') as f:\n",
    "    table.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 638,
   "id": "feab609d-1b23-4101-a709-1fc9c6ebad40",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = metadata.set_index(\"sample_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 639,
   "id": "314c7757-dce6-4321-884b-bfda2b90d566",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = biom.load_table(\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/table.biom\")\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "sid = table.ids(axis=\"sample\")\n",
    "n = 1\n",
    "for train_idx, test_idx in kf.split(sid):\n",
    "    table_train = table.filter(sid[train_idx], axis=\"sample\", inplace=False)\n",
    "    table_train.remove_empty()\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/data/train_{n}.biom\", 'w') as f:\n",
    "        table_train.to_hdf5(f, 'example')\n",
    "    \n",
    "    table_test = table.filter(sid[test_idx], axis=\"sample\", inplace=False)\n",
    "    table_test.remove_empty()\n",
    "    with biom.util.biom_open(f\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/data/test_{n}.biom\", 'w') as f:\n",
    "        table_test.to_hdf5(f, 'example')\n",
    "\n",
    "    table_train = table_train.to_dataframe().T\n",
    "    table_train.loc[:, \"Group\"] = metadata.loc[table_train.index.values, \"group\"]\n",
    "    table_train.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/data/train_{n}.csv\")\n",
    "    table_test = table_test.to_dataframe().T\n",
    "    table_test.loc[:, \"Group\"] = metadata.loc[table_test.index.values, \"group\"]\n",
    "    table_test.to_csv(f\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/data/test_{n}.csv\")\n",
    "        \n",
    "    n = n + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 644,
   "id": "5caa3c6e-8e23-4c3a-8768-d17c922be253",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_1 = TreeNode.read(\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/phylogeny.nwk\")\n",
    "tree_2 = TreeNode.read(\"/home/dongbiao/GCN/data/synthetic_data/phylogeny.nwk\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 645,
   "id": "be163059-6725-4abe-aade-7d61e7a04dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_1_distance = tree_1.tip_tip_distances()\n",
    "tree_2_distance = tree_2.tip_tip_distances()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 647,
   "id": "9cc38537-15db-457b-a4cb-e81ca024dd0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.     , 0.21668, 0.2518 , ..., 0.38665, 0.37041, 0.35561],\n",
       "       [0.21668, 0.     , 0.22174, ..., 0.41933, 0.40309, 0.38829],\n",
       "       [0.2518 , 0.22174, 0.     , ..., 0.45445, 0.43821, 0.42341],\n",
       "       ...,\n",
       "       [0.38665, 0.41933, 0.45445, ..., 0.     , 0.2272 , 0.24656],\n",
       "       [0.37041, 0.40309, 0.43821, ..., 0.2272 , 0.     , 0.23032],\n",
       "       [0.35561, 0.38829, 0.42341, ..., 0.24656, 0.23032, 0.     ]])"
      ]
     },
     "execution_count": 647,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tree_1_distance.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 648,
   "id": "2d295e30-dff1-4c6d-bc7f-d1ceb5638c1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  0.      ,  25.260833, 117.193324, ..., 332.71476 , 310.231344,\n",
       "        294.66027 ],\n",
       "       [ 25.260833,   0.      , 120.236223, ..., 335.757659, 313.274243,\n",
       "        297.703169],\n",
       "       [117.193324, 120.236223,   0.      , ..., 308.432318, 285.948902,\n",
       "        270.377828],\n",
       "       ...,\n",
       "       [332.71476 , 335.757659, 308.432318, ...,   0.      , 199.988276,\n",
       "        372.112816],\n",
       "       [310.231344, 313.274243, 285.948902, ..., 199.988276,   0.      ,\n",
       "        349.6294  ],\n",
       "       [294.66027 , 297.703169, 270.377828, ..., 372.112816, 349.6294  ,\n",
       "          0.      ]])"
      ]
     },
     "execution_count": 648,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tree_2_distance.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 655,
   "id": "fead442b-4067-4a55-a441-2323fa05443f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gaussian_kernel(d, rho=2):\n",
    "    \"\"\"\n",
    "    weight = exp(-2 * ρ * d^2)\n",
    "    \"\"\"\n",
    "    return np.exp(-2 * rho * d ** 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 665,
   "id": "e73d0a0a-e1a8-40db-9aee-39c5fec71c03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9999840001279994"
      ]
     },
     "execution_count": 665,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gaussian_kernel(0.4, rho=5e-5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a0ab02a3-8dae-4e67-9f25-e45337684929",
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_tree_for_gnn(input_path, output_path):\n",
    "    tree = TreeNode.read(input_path)\n",
    "    max_depth = 0.0\n",
    "    for tip in tree.tips():\n",
    "        dist = tip.accumulate_to_ancestor(tree)\n",
    "        if dist > max_depth:\n",
    "            max_depth = dist\n",
    "\n",
    "    count = 0\n",
    "    for node in tree.traverse():\n",
    "        if node.length is not None:\n",
    "            node.length = node.length / max_depth\n",
    "            count += 1\n",
    "\n",
    "    tips = list(tree.tips())\n",
    "    if len(tips) >= 2:\n",
    "        d = tips[0].distance(tips[1])\n",
    "\n",
    "    tree.write(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 867,
   "id": "969afdb1-55e5-4c55-a0f3-ea8fefd5329c",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/GCN/data/synthetic_data/phylogeny.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/GCN/data/synthetic_data/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 805,
   "id": "a3f4ff53-02d5-453a-8c91-99611d774799",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/GCN/data/IBD/phylogeny.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/GCN/data/IBD/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 834,
   "id": "ada7eb9a-13f5-4765-851a-bff4c7f40b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/GCN/data/CRC/phylogeny.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/GCN/data/CRC/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 854,
   "id": "664dee6b-99af-456c-a995-a75dfb45fd15",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/GCN/data/dietary_fiber/phylogeny.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/GCN/data/dietary_fiber/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1055,
   "id": "1ff096b6-5d0f-4c20-907f-3fc9b0a1d363",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/GCN/data/OSCC_16S/phylogeny.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/GCN/data/OSCC_16S/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "fbfec0f3-301e-4cfe-9267-4fc37a2f71fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/16S_CRC.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "03651046-6d15-4697-bd70-5115cc46d8f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/GCN/data/dietary_fiber/phylogeny.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/GCN/data/dietary_fiber/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1019,
   "id": "31ac1d4e-e909-4caa-8121-d74f337e0e9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/phylogeny.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1044,
   "id": "7ca58526-0ac6-49f7-8611-f2163eee345b",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/phylogeny.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "cfe22e63-7259-4d15-9774-980d1f3bc346",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tree_file = \"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/phylogeny.nwk\" \n",
    "output_tree_file = \"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/phylogeny_normalize.nwk\"\n",
    "normalize_tree_for_gnn(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "203895f4-449c-4e56-8331-3c5b45cb3429",
   "metadata": {},
   "source": [
    "### Deepphylo Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "3c5795d1-df77-4d73-a54d-fcf9513f9b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "reducer = PCA(n_components=32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "f9065e3a-7700-4cab-bddd-7fde22d53706",
   "metadata": {},
   "outputs": [],
   "source": [
    "### synthetic data\n",
    "tree = TreeNode.read(\"/home/dongbiao/software/Phylo-Spec/data/Synthetic_Dataset_1/phylogeny.nwk\")\n",
    "dm = tree.tip_tip_distances()\n",
    "dm = pd.DataFrame(data=dm.data, index=dm.ids, columns=dm.ids)\n",
    "phy_embedding = reducer.fit_transform(dm.values)\n",
    "phy_embedding = pd.DataFrame(data=phy_embedding, index=dm.index.values)\n",
    "phy_embedding.to_csv(\"/home/dongbiao/GCN/data/synthetic_data/PCA_32.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "d116db62-d4bf-4248-a0b9-e1b6e432e3c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### IBD 16S\n",
    "tree = TreeNode.read(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_IBD/phylogeny_normalize.nwk\")\n",
    "dm = tree.tip_tip_distances()\n",
    "dm = pd.DataFrame(data=dm.data, index=dm.ids, columns=dm.ids)\n",
    "phy_embedding = reducer.fit_transform(dm.values)\n",
    "phy_embedding = pd.DataFrame(data=phy_embedding, index=dm.index.values)\n",
    "phy_embedding.to_csv(\"/home/dongbiao/GCN/data/IBD_16S/PCA_32.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "3aa12684-b211-43fa-baeb-74319d376cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "### CRC 16S\n",
    "tree = TreeNode.read(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_16S_CRC/phylogeny_normalize.nwk\")\n",
    "dm = tree.tip_tip_distances()\n",
    "dm = pd.DataFrame(data=dm.data, index=dm.ids, columns=dm.ids)\n",
    "phy_embedding = reducer.fit_transform(dm.values)\n",
    "phy_embedding = pd.DataFrame(data=phy_embedding, index=dm.index.values)\n",
    "phy_embedding.to_csv(\"/home/dongbiao/GCN/data/CRC_16S//PCA_32.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "1db0d291-63f8-4a9a-a54c-31e1efcbdd86",
   "metadata": {},
   "outputs": [],
   "source": [
    "### dietary fiber 16S\n",
    "tree = TreeNode.read(\"/home/dongbiao/GCN/data/dietary_fiber/phylogeny_normalize.nwk\")\n",
    "dm = tree.tip_tip_distances()\n",
    "dm = pd.DataFrame(data=dm.data, index=dm.ids, columns=dm.ids)\n",
    "phy_embedding = reducer.fit_transform(dm.values)\n",
    "phy_embedding = pd.DataFrame(data=phy_embedding, index=dm.index.values)\n",
    "phy_embedding.to_csv(\"/home/dongbiao/GCN/data/dietary_fiber/PCA_32.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "80600f11-d11b-4d90-a721-dd1f0a125d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "### CRC WGC\n",
    "tree = TreeNode.read(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_CRC/phylogeny_normalize.nwk\")\n",
    "dm = tree.tip_tip_distances()\n",
    "dm = pd.DataFrame(data=dm.data, index=dm.ids, columns=dm.ids)\n",
    "phy_embedding = reducer.fit_transform(dm.values)\n",
    "phy_embedding = pd.DataFrame(data=phy_embedding, index=dm.index.values)\n",
    "phy_embedding.to_csv(\"/home/dongbiao/GCN/data/CRC_WGS/PCA_32.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "4dd5fae4-fb49-4d76-94c2-2c73dcf70f9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### T2D WGS\n",
    "tree = TreeNode.read(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_WGS_T2D/phylogeny_normalize.nwk\")\n",
    "dm = tree.tip_tip_distances()\n",
    "dm = pd.DataFrame(data=dm.data, index=dm.ids, columns=dm.ids)\n",
    "phy_embedding = reducer.fit_transform(dm.values)\n",
    "phy_embedding = pd.DataFrame(data=phy_embedding, index=dm.index.values)\n",
    "phy_embedding.to_csv(\"/home/dongbiao/GCN/data/T2D_WGS/PCA_32.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "66296dc7-5451-46f9-bae8-81c651420e2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Multi-classification 16S\n",
    "tree = TreeNode.read(\"/home/dongbiao/software/Phylo-Spec/data/Real_Dateset_Multi-classification/phylogeny_normalize.nwk\")\n",
    "dm = tree.tip_tip_distances()\n",
    "dm = pd.DataFrame(data=dm.data, index=dm.ids, columns=dm.ids)\n",
    "phy_embedding = reducer.fit_transform(dm.values)\n",
    "phy_embedding = pd.DataFrame(data=phy_embedding, index=dm.index.values)\n",
    "phy_embedding.to_csv(\"/home/dongbiao/GCN/data/Multi_classification/PCA_32.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f452de2c-da85-4109-a204-edc68b6fa91a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
